### PR TITLE
Add abs call over default error estimates.

### PR DESCRIPTION
**Pull request by grimmmyshini**
This PR introduces the following changes:
- Build a `std::abs` call over the error estimates from the Taylor approx model.
- Two helper functions for users to build their own function calls.

### CODE REVIEW

File: include/clad/Differentiator/EstimationModel.h [hunk 1]
```diff
@@ -48,6 +48,39 @@ namespace clad {
...
+    /// currently does not support nested namespaces.
+    ///
+    /// \return The unresolved lookup expression that can be used to build
+    /// actual function call. This cannot be used directly to emit code, use
+    /// \c BuildFucntionCallExpr instead.
+    clang::Expr* GetFunctionCall(std::string funcName,
+                                 std::string nmspace = "");
+
+    /// Helper to build a function call expression.
+    /// \note This function does not check if the parameters result in correct
```
Comment (vgvassilev) [inline include/clad/Differentiator/EstimationModel.h:67]: Did you actually try that? The code you wrote on `GetFunctionCall` seems to be pretty generic to handle overloads. Calling `ActOnCallExpr` with `UnresolvedLookupExpr` does the overload resolution. If the current code doesn't do that, I'd prefer to allow this behavior as I am not very concerned about the performance overhead of overload resolution.

Comment (grimmmyshini) [reply include/clad/Differentiator/EstimationModel.h:67]: I am not sure if I fully understand your question/comment. I just removed the explicit `noOverloadExists` check because it is very very unlikely that the user is passing incorrect parameters for overload resolution. And anyway it will be verified when we build the call expression/during compilation, so no need for an extra step that will be done anyway at a later point in time.

Comment (vgvassilev) [reply include/clad/Differentiator/EstimationModel.h:67]: @grimmmyshini, ok, I think we need to adjust the `\note` in the comment. My claim is that the implementation will do the overload resolution. Can you check?

Comment (grimmmyshini) [reply include/clad/Differentiator/EstimationModel.h:67]: Oh yes, it does, why do we have the 'NoOverloadExists' function then?

Comment (parth-07) [reply include/clad/Differentiator/EstimationModel.h]: `NoOverloadExists` function is used to ensure that an overload exists for a given unresolved expression and an argument list. It does not determine the overload itself.

Comment (grimmmyshini) [reply include/clad/Differentiator/EstimationModel.h]: It is not really required but I suppose you need it so that you can fail gracefully. In all other cases using an ActOnCallExpr is sufficient.

Comment (parth-07) [reply include/clad/Differentiator/EstimationModel.h]: > It is not really required but I suppose you need it so that you can fail gracefully.

Yes, exactly.

File: lib/Differentiator/EstimationModel.cpp [hunk 2]
```diff
@@ -49,20 +52,64 @@ namespace clad {
   Expr* FPErrorEstimationModel::SetError(VarDecl* declStmt) { 
     return nullptr; 
   }
+  
+  Expr* FPErrorEstimationModel::GetFunctionCall(std::string funcName,
```
Comment (vgvassilev) [inline lib/Differentiator/EstimationModel.cpp:56]: Why not merging this function with `BuildFunctionCallExpr`. Is there a case in the codebase where we need to build an `UnresolvedLookup` without then calling `BuildFunctionCallExpr`?

File: lib/Differentiator/EstimationModel.cpp [hunk 3]
```diff
@@ -49,20 +52,64 @@ namespace clad {
...
+    Expr* UnresolvedLookup = nullptr;
+    if (!R.empty())
+      UnresolvedLookup =
+          m_Sema.BuildDeclarationNameExpr(SS, R, /*ADL=*/false).get();
+    return UnresolvedLookup;
+  }
+
+  Expr* FPErrorEstimationModel::BuildFunctionCallExpr(
+      Expr* lookupRes, llvm::SmallVectorImpl<Expr*>& callArgs) {
+    if (!lookupRes)
```
Comment (vgvassilev) [inline lib/Differentiator/EstimationModel.cpp:80]: I'd prefer to turn this into an `assert` as this is a programmatic error if we don't reach here.

File: include/clad/Differentiator/EstimationModel.h [hunk 4]
```diff
@@ -129,11 +157,15 @@ namespace clad {
 
   /// Example class for taylor series approximation based error estimation.
   class TaylorApprox : public FPErrorEstimationModel {
+    /// A variable to keep track of the function lookup expression for the
+    /// \c std::abs call
+    clang::Expr* absCallExpr = nullptr;
```
Comment (vgvassilev) [inline include/clad/Differentiator/EstimationModel.h:162]:
```suggestion
    clang::Expr* m_absCallExpr = nullptr;
```

**Review by vgvassilev — COMMENTED** [review]
Thanks for working on this. Here are some suggestions.

File: lib/Differentiator/EstimationModel.cpp (hunk 2 above)
Comment (grimmmyshini) [reply lib/Differentiator/EstimationModel.cpp:56]: Can you rephrase this? I don't think I understand...

Comment (vgvassilev) [reply lib/Differentiator/EstimationModel.cpp:56]: IIUC, the optimization you have in mind is reusing the `UnresolvedLookupExpr` built by this function. I am saying that we should not do this, as we should have an `UnresolvedLookupExpr` per CallExpr.

Comment (grimmmyshini) [reply lib/Differentiator/EstimationModel.cpp:56]: Okay, so each call expr should have its own unresolved lookup expr?

Comment (vgvassilev) [reply lib/Differentiator/EstimationModel.cpp:56]: Correct!

**Review by vgvassilev — APPROVED** [review]
LGTM!
